cartoframes/contrib/function.py

```diff
@@ -0,0 +1,31 @@
+def add_function(func, func_name, func_description, context):
+    import dill
+    import pandas as pd
+    pkl_func = dill.dumps(func)
+    new_func = pd.DataFrame([[func_name, pkl_func, func_description, None]],
+                          columns=['name', 'function', 'description', 'the_geom'])
+    try:
+        existing_func = context.read('functions')
+    except:
+        existing_func = pd.DataFrame(columns=['name', 'function', 'description', 'the_geom'])
+    assert not func_name in existing_func.name.values, 'Function Name Duplication Error: Please Rename The Function'
```

Comment:
What about just reading the row that matches the function name instead of pulling the entire dataframe/table?

Reply:
good idea, e.g. `cc.query(f'select * from functions where name = {func_name}')`. But in the end I have to overwrite the whole `functions` table to my account. Wondering if there's an option that allows us to only append new rows to the existing table. I checked Import API and didn't find a good solution.

Reply:
Using SQL you can do an INSERT, or better yet an UPSERT. It might be useful to have a hash of the functions and only INSERT if the the name does not exist, and UPDATE if the name is the same but the hash changes. Another idea is to have _versions_ so that you don't have to worry about overwrites. The functions table could have version numbers and only the latest version is flagged as 'active'? Just throwing some ideas out there :)

Reply:
That's awesome!! I'll try it. 
Actually, my original idea is to create a column called `project`. For different projects, users could save functions with the same name (but maybe different versions, `v_project1`, `v_project1`), which allows them to save both universal functions or very customized only for a specific project). And every time they want to run (or rerun) the data science analytical pipelines, they just need to load all functions using `project` name and implement tools like (Python) `sklearn.pipeline.Pipeline` or (R) `%>%`. I think this will make notebooks/scripts neat and tidy especially for those non ad-hoc analyses.